docs/architecture.md

```diff
@@ -8,209 +8,175 @@
--- `REJECTED`.
-CREATE TABLE deal (
-  aggregate TEXT PRIMARY KEY NOT NULL REFERENCES aggregate(link),
-  status DEAL_STATUS DEFAULT 'PENDING',
-  -- if status is an error this may contain details about the error e.g. json containing
-  -- piece CIDs that were invalid.
-  detail JSONB,
-  -- Time when aggregate was send to spade-proxy
-  inserted TIMESTAMP WITH TIME ZONE DEFAULT timezone('utc'::text, now()) NOT NULL,
-  -- Time when aggregate was signed
-  signed TIMESTAMP WITH TIME ZONE,
-  -- Time when aggregate was processed (became APPROVED or REJECTED)
-  processed TIMESTAMP WITH TIME ZONE
-);
-
-CREATE INDEX deal_inserted_idx ON deal (inserted);
-CREATE INDEX deal_signed_idx ON deal (signed);
-CREATE INDEX deal_signed_idx ON deal (signed);
-CREATE INDEX deal_status_idx ON deal (status);
-
-CREATE TYPE DEAL_STATUS AS ENUM
-(
-  'PENDING',
-  'SIGNED',
-  'APPROVED',
-  'REJECTED'
-);
-
--- View of aggregates to be submitted to spade, that is all aggregates that we do not
--- have deal records for.
-CREATE VIEW aggregate_queue AS
-  SELECT aggregate.*
-  FROM aggregate
-  LEFT OUTER JOIN deal ON aggregate.link = deal.aggregate
-  WHERE deal.aggregate IS NULL;
-
--- View for deals pending.
-CREATE VIEW deal_pending AS
-  SELECT *
-  FROM deal
-  WHERE status = "PENDING"
-  ORDER BY inserted;
-
--- View for deals approved by storage providers.
-CREATE VIEW deal_approved AS
-  SELECT *
-  FROM deal
-  WHERE status = "APPROVED"
-  ORDER BY processed;
-
--- View for deals rejected by storage providers.
-CREATE VIEW deal_rejected AS
-  SELECT *
-  FROM deal
-  WHERE status = "REJECTED"
-  ORDER BY processed;
+## Processor Stack
+
+When a `piece` is posted into the w3filecoin pipeline, its journey starts by getting queued to be aggregated into a larger `piece` that will be offered to Storage Providers, i.e. an aggregate. The `Queue stack` consists of a **multiple queue system**, where the individual pieces will be buffered together in several stages until they are ready to form an aggregate (**32 GiB** piece, or close to this size).
+
+This design is built on top of the following assumptions:
+- Maximum SQS batch size for standard queue is **10_000**
+- Maximum SQS batch size for FIFO queue is **10**
+- SQS FIFO queue garantees **exactly-once** processing
+- Maximum number of pieces for a **32 GiB** aggregate is **262_144**
+
+The `piece-queue` is the first queue in this system and is a standard SQS queue. It buffers individual pieces until a batch of **10_000** is ready. Once this batch is ready, a SQS consumer will try to create smaller piece aggregates named Ferries. A ferry is a `dag-cbor` encoded data structure that contains a set of pieces that form a partial aggregate. This data structure will be stored so that only its CID is sent in SQS Messages.
+
+```typescript
+interface Ferry {
+  // Pieces inside the ferry
+  pieces: FerryPiece[],
+  // Aggregate composed by pieces in ferry enabling us to derive its size
+  aggregate: FerryPiece,
```

Comment:
I added the aggregate because getting to know its size right away could be helpful if we ever wanted to increase batch size. But I think you are right, we can get it out for now